### ISSUE TITLE
numpy and scipy qr have different mode names.

### ISSUE THREAD

**Issue by charris** [scipy/scipy#4818]
In numpy >= 1.8 the modes are
```
mode : {'reduced', 'complete', 'r', 'raw', 'full', 'economic'}, optional
    If K = min(M, N), then

    'reduced'  : returns q, r with dimensions (M, K), (K, N) (default)
    'complete' : returns q, r with dimensions (M, M), (M, N)
    'r'        : returns r only with dimensions (K, N)
    'raw'      : returns h, tau with dimensions (N, M), (K,)
    'full'     : alias of 'reduced', deprecated
    'economic' : returns h from 'raw', deprecated.

    The options 'reduced', 'complete, and 'raw' are new in numpy 1.8,
    see the notes for more information. The default is 'reduced' and to
    maintain backward compatibility with earlier versions of numpy both
    it and the old default 'full' can be omitted. Note that array h
    returned in 'raw' mode is transposed for calling Fortran. The
    'economic' mode is deprecated.  The modes 'full' and 'economic' may
    be passed using only the first letter for backwards compatibility,
    but all others must be spelled out. See the Notes for more
    explanation.
```
Whereas scipy has
```
mode : {'full', 'r', 'economic', 'raw'}, optional
    Determines what information is to be returned: either both Q and R
    ('full', default), only R ('r') or both Q and R but computed in
    economy-size ('economic', see Notes). The final option 'raw'
    (added in Scipy 0.11) makes the function return two matrices
    (Q, TAU) in the internal format used by LAPACK.
```
Note that `full` has different meanings in the two cases, which was one reason for the addition of the modes `complete` and `reduced` in numpy, plus the fact that numpy did not have a `complete` mode ('full' in scipy).

**Comment by rgommers** [comment]
Did we discuss before what needed changing in scipy?

**Comment by h-vetinari** [comment]
[Status](https://numpy.org/devdocs/reference/generated/numpy.linalg.qr.html) numpy 1.22:
```
mode : {‘reduced’, ‘complete’, ‘r’, ‘raw’}
```
[Status](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.qr.html) scipy 1.8:
```
mode : {‘full’, ‘r’, ‘economic’, ‘raw’},
```

**Comment by lucascolley** [comment]
Nothing has changed in NumPy and SciPy since the last status update. We also have [the array API standard](https://data-apis.org/array-api/draft/extensions/generated/array_api.linalg.qr.html#array_api.linalg.qr) now, which has:

```
mode (Literal['reduced', 'complete'])
```

so the deprecation should happen on SciPy's side, if it is deemed appropriate.

**Comment by ilayn** [comment]
Seems like array API went a bit too fast with these ones to be honest (missing `pivoting` which is needed for RRQR and so on). 
But why are they not in the linalg extension? 

Also we don't need to do anything here since this is baked into the array namespace. Hence the array owners should implement these. 

Our signature has more important issues like exposed `lwork`, changing output arg numbers and weird `mode` names.

**Comment by lucascolley** [comment]
> Also we don't need to do anything here since this is baked into the array namespace. Hence the array owners should implement these.

Do you envision `scipy.linalg.qr` eventually being removed from SciPy then? Not quite sure if I've understood you correctly. But yeah, that would close this issue 😅

**Comment by ilayn** [comment]
No I mean array API decided to implement their own in their own namespace so what's the point of calling scipy if the array owner has their own native implementation?

**Comment by ilayn** [comment]
Wait I think this is in the linalg extension, the left menu is not displaying correctly.

**Comment by lucascolley** [comment]
right, so if `linalg.qr` is going to remain in both NumPy and SciPy for the foreseeable future, it would be nice to settle on 'reduced' and 'complete' rather than 'full' and 'economic'. I don't know how to judge whether the deprecation churn would be worth it, though.

**Comment by charris** [comment]
There was an extended discussion about the QR keyword names back when they were changed. The main difference between scipy and numpy, apart from the names, is the default.  There has also been some discussion of adding a rank preserving qr (column pivoting) version, but it didn't reach any conclusions IIRC.

**Comment by rgommers** [comment]
This is pretty much unrelated to the array API standard. Rather, it's one of a handful remaining functions that are both in `numpy` and `scipy`, and resolving the differences would be nice in general. We did so for all FFT functionality; for `linalg` there's a few issues like this one left.

> I don't know how to judge whether the deprecation churn would be worth it, though.

Agreed, it requires careful consideration, unclear to me that this is worth it. It's a nice to have for consistency I believe, not a must-do.

> No I mean array API decided to implement their own in their own namespace so what's the point of calling scipy if the array owner has their own native implementation?

Agreed. And if `scipy.linalg.qr` wants to defer to (e.g.) `torch.linalg.qr`, that would be an internal implementation detail.

> Our signature has more important issues like exposed `lwork`, changing output arg numbers and weird `mode` names.

Agreed. `overwrite_a` is also not pretty.

**Comment by ev-br** [comment]
So as long as numpy's "complete" option is the same as scipy's "full", how about scipy adding "complete" as an alias for "full"? Same for "reduced" aliasing "economic".
